0.15 Release Notes: Cached One-shot Systems (#1813)
bevyengine/bevy-website: release-content/0.15/release-notes/14920_Add_cached_run_system_API.md

# Cached one-shot systems

> **Bevy 0.15** introduces a new "cached" API for running one-shot systems:

Reviewer: ```suggestion
**Bevy 0.15** introduces a convenient new "cached" API for running one-shot systems:
```

```rust

> // Run a system without registering it first:

Author: ```suggestion
// Run a system without manually registering it first:
```

commands.run_system_cached(foo);

> commands.run_system_cached_with(bar, 42);

Reviewer: ```suggestion
// You can pass input into one-shot systems!
commands.run_system_cached_with(bar, 42);
```

Author: This might read like passing inputs is a new feature. It does help explain this line though, since passing input into a system is less common. Maybe the `run_system_cached_with` line should just be removed?

Reviewer: Yeah, let's cut it.

> // Reuse a registered system automatically:

Reviewer: I don't think talking about system registration is useful.

Author: As in, don't mention `register_system_cached` up here? I know `register_system_cached` will be used less frequently than `run_system_cached`, but I figured that showing how it gives the same ID when called multiple times was a good demonstration of what the API does differently.

Reviewer: Ah, I see your point now. Right, so I think that this is best left as a demonstration of the technical details below.

Author: Moved it later and added contrast.

let id1 = world.register_system_cached(quux);
let id2 = world.register_system_cached(quux);
assert_eq!(id1, id2);
```

> To better understand why this new API is useful, let's take a look at what systems really are behind the curtains:

Reviewer: IMO we should show the existing API to demonstrate how much nicer this is, and *then* get into the weeds about the technical details.

Author: Made the first code block compare old vs. new API.

So, for example, suppose you write some code that registers and runs a system whenever a button is pressed. Unfortunately, you now have a memory leak: whenever the button is pressed, a new copy of the system will spawn with its own system state. A common pattern that fixes this leak is to cache the `SystemId` in a `Local` or `Resource` and then only register the system the first time the button is pressed. This is exactly what the new "cached" API does for you! When you `run_system_cached`, the `SystemId` will be saved in a generic `CachedSystemId<S: System>` resource to be reused later. You can `unregister_system_cached` to remove the system and its `CachedSystemId`.

## Limitations

> This approach only works if we can guarantee that different systems are never cached under the same `CachedSystemId<S>`. In other words, there should be no more than 1 distinct system of type `S`. This is true when `size_of::<S>() == 0` (aka `S` is a ZST), which is almost always true in practice. Thus, to enforce correctness, the new API will give you a compile-time error if you pass in a non-ZST function like a function pointer or a capturing closure.

Reviewer: ```suggestion
This approach only works if we can guarantee that different systems are never cached under the same `CachedSystemId<S>`. In other words, there should be no more than 1 distinct system of type `S`. This is true when `size_of::<S>() == 0`, which is almost always true in practice. Thus, to enforce correctness, the new API will give you a compile-time error if you try to use a non-zero-sized function (like a function pointer or a capturing closure).
```

## Comparison to `run_system_once`

> The approach taken by `run_system_once` is to set up a system, run it once, and then tear it down. This means that `run_system_once` _does_ avoid the memory leak problem. However, any system parameters like `Local` and `EventReader` that rely on persistent state between runs will be broken; and any system parameters like `Query` that rely on cached computations to improve performance will have to rebuild their cache each time, which can be costly. As a consequence, `run_system_once` is only recommended for diagnostic use (e.g. unit tests), and `run_system` or `run_system_cached` should be preferred for "real" code.

Reviewer: This is valuable, but I think we can get a better framing by starting off with "there were two ways, now there's three". Move this up?

Author: Moved this before the Limitations section and added "there were two ways, now there's three" to the start. Not 100% confident I've understood the feedback correctly though.

Reviewer: I was intending to put this right up at the beginning with `run_system`, `run_system_cached` and `run_system_once`. Your approach is nice too though, so I'm merging.
